pandas/core/aggregation.py

```diff
@@ -415,67 +423,94 @@ def transform(
-
-        results = {}
-        for name, how in func.items():
-            colg = obj._gotitem(name, ndim=1)
-            try:
-                results[name] = transform(colg, how, 0, *args, **kwargs)
-            except Exception as e:
-                if str(e) == "Function did not transform":
-                    raise e
-
-        # combine results
-        if len(results) == 0:
+    if is_dict_like(func):
+        func = cast(Dict[Label, Union[AggFuncTypeBase, List[AggFuncTypeBase]]], func)
+        result = transform_dict_like(obj, func, *args, **kwargs)
+    else:
+        func = cast(AggFuncTypeBase, func)
+        try:
+            result = transform_str_or_callable(obj, func, *args, **kwargs)
+        except Exception:
             raise ValueError("Transform function failed")
-        return concat(results, axis=1)
-
-    # func is either str or callable
-    try:
-        if isinstance(func, str):
-            result = obj._try_aggregate_string_function(func, *args, **kwargs)
-        else:
-            f = obj._get_cython_func(func)
-            if f and not args and not kwargs:
-                result = getattr(obj, f)()
-            else:
-                try:
-                    result = obj.apply(func, args=args, **kwargs)
-                except Exception:
-                    result = func(obj, *args, **kwargs)
-    except Exception:
-        raise ValueError("Transform function failed")
 
     # Functions that transform may return empty Series/DataFrame
     # when the dtype is not appropriate
     if isinstance(result, (ABCSeries, ABCDataFrame)) and result.empty:
         raise ValueError("Transform function failed")
+
     if not isinstance(result, (ABCSeries, ABCDataFrame)) or not result.index.equals(
         obj.index
     ):
         raise ValueError("Function did not transform")
 
     return result
+
+
+def transform_dict_like(
+    obj: FrameOrSeries,
+    func: Dict[Label, Union[AggFuncTypeBase, List[AggFuncTypeBase]]],
+    *args,
+    **kwargs,
+):
+    """
+    Compute transform in the case of a dict-like func
+    """
+    from pandas.core.reshape.concat import concat
+
+    if obj.ndim != 1:
+        # Check for missing columns on a frame
+        cols = sorted(set(func.keys()) - set(obj.columns))
+        if len(cols) > 0:
+            raise SpecificationError(f"Column(s) {cols} do not exist")
+
+    if any(isinstance(v, dict) for v in func.values()):
+        # GH 15931 - deprecation of renaming keys
+        raise SpecificationError("nested renamer is not supported")
+
+    results: Dict[Label, FrameOrSeriesUnion] = {}
+    for name, how in func.items():
+        colg = obj._gotitem(name, ndim=1)
+        try:
+            results[name] = transform(colg, how, 0, *args, **kwargs)
+        except Exception as e:
+            if str(e) == "Function did not transform":
+                raise e
+
+    # combine results
+    if len(results) == 0:
+        raise ValueError("Transform function failed")
```

Comment:
is this tested? is this new?

Reply:
i think you should just let concat work and bubble up if needed

Reply:
Not new, this is hit in `test.frame.apply.test_frame_transform.test_transform_bad_dtype`. Removing these two lines results in the ValueError "No objects to concatenate" rather than "Transform function failed". I'm okay with either, slight preference for the current behavior (which is the same as this PR). Let me know if you prefer "No objects to concatenate" and can change.

Reply:
hmm yeah I think we should change this, maybe to something a bit more userfriendly like
`the input transform did not contain any transform functions`

Reply:
Are you thinking of the case where an aggregator is used? If we detect the function not transforming, we'll raise here:

```
try:
    results[name] = transform(colg, how, 0, *args, **kwargs)
except Exception as e:
    if str(e) == "Function did not transform":
        raise e
```

before we get to this point. The raising of the code highlighted above should only occur if all the key/value pairs of the dictionary entirely failed - e.g. trying to take a product of strings or `lambda x: raise ValueError`.

Reply:
ok, no this is the case of an empty list or dict right? (of functions that are supplied for aggregation), IOW its user visible

Reply:
Ah, I see! I did not realize this line would be the one that raised for an empty list or dict. I've added that check before getting to this section of the code with the message "no results" (this is the message from 1.1 - but can change), along with tests for this.